### PR TITLE
Initialize authentication popup modal only when needed

### LINKED ISSUE #30672: [Issue] Initialize authentication popup modal only when needed
This issue is automatically created based on existing pull request: magento/magento2#30647: Initialize authentication popup modal only when needed

---------

### Description (*)
This PR is somehow an evolution of the https://github.com/magento/magento2/pull/27686 in which I have found that it is not possible to completely skip initialization of authentication popup. That said, by adding this additional check, we are able to save a lot of CPU time for all of the shops that have guest checkout enabled.

Here are the profiles of before:
auth-popup-before
and after:
auth-popup-after

### Manual testing scenarios (*)

1. `setModalElement` function is no longer called when guest checkout is enabled.
2. Authentication popup still shows properly when guest user tries to go to the checkout but guest checkout is disabled.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### PR DESCRIPTION
**Pull request by krzksz**
### Description (*)
This PR is somehow an evolution of the https://github.com/magento/magento2/pull/27686 in which I have found that it is not possible to completely skip initialization of authentication popup. That said, by adding this additional check, we are able to save a lot of CPU time for all of the shops that have guest checkout enabled.

### Related Pull Requests
https://github.com/magento/inventory/pull/3261
https://github.com/magento/partners-magento2ee/pull/428

Here are the profiles of before:
auth-popup-before
and after:
auth-popup-after

### Manual testing scenarios (*)

1. `setModalElement` function is no longer called when guest checkout is enabled.
2. Authentication popup still shows properly when guest user tries to go to the checkout but guest checkout is disabled.

### Contribution checklist (*)
 - [ ] Pull request has a meaningful description of its purpose
 - [ ] All commits are accompanied by meaningful commit messages
 - [ ] All new or changed code is covered with unit/integration tests (if applicable)
 - [ ] All automated tests passed successfully (all builds are green)

### Resolved issues:
1. [x] resolves magento/magento2#30672: Initialize authentication popup modal only when needed

### CODE REVIEW

**Comment by mrtuvn** [comment]
Nice works. This pr will make magento more become better place

**Comment by krzksz** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
Can you add minimal dependency js like closed PR 27686. Seem current authen-popup require seem a lot another components

**Review by ihor-sviziev — APPROVED** [review]
Seems like your changes causing regression. Could you review test failures and adjust your code accordingly?

**Review by ihor-sviziev — CHANGES_REQUESTED** [review]
Seems like your changes causing regression. Could you review test failures and adjust your code accordingly?

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8387](https://jira.corp.magento.com/browse/ENGCOM-8387) has been created to process this Pull Request
:eight_spoked_asterisk: @ihor-sviziev, could you please add one of the following labels to the Pull Request?

| Label | Description |
|--------------------------------|-----------------------------------------------------------------|
| ```Auto-Tests: Covered``` | All changes in Pull Request is covered by auto-tests |
| ```Auto-Tests: Not Covered``` | Changes in Pull Request requires coverage by auto-tests |
| ```Auto-Tests: Not Required``` | Changes in Pull Request does not require coverage by auto-tests |

**Comment by sidolov** [comment]
@magento create issue

**Comment by ihor-sviziev** [comment]
@magento run all tests

**Comment by krzksz** [comment]
@magento run Functional Tests CE,Static Tests,Unit Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE, Unit Tests

**Comment by krzksz** [comment]
It seems we are on a good path. There are still some tests which try to login customer using hidden authentication popup instead of actual forms because of duplicated ID, but I'll try to adjust the selectors to fix that.

**Comment by ihor-sviziev** [comment]
@krzksz that's actually really good catch!

**Comment by engcom-Charlie** [comment]
@krzksz could you please fix failed tests?
Thank you.

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

File: app/code/Magento/Checkout/Test/Mftf/Section/CheckoutShippingGuestInfoSection.xml [hunk 1]
```diff
@@ -9,7 +9,7 @@
 <sections xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
         xsi:noNamespaceSchemaLocation="urn:magento:mftf:Page/etc/SectionObject.xsd">
     <section name="CheckoutShippingGuestInfoSection">
-        <element name="email" type="input" selector="#customer-email"/>
+        <element name="email" type="input" selector="#customer-email-fieldset #customer-email"/>
```
Comment (ihor-sviziev) [inline app/code/Magento/Checkout/Test/Mftf/Section/CheckoutShippingGuestInfoSection.xml]: that's pretty stupid to looking for ID in the ID.
can we use some class?

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests B2B

**Review by ihor-sviziev — APPROVED** [review]
Finally, everything is green! 🎉
✔ Approved

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8387](https://jira.corp.magento.com/browse/ENGCOM-8387) has been created to process this Pull Request

**Comment by engcom-Charlie** [comment]
@ihor-sviziev Looks like changes are covered by the js unit test. Can we set the label `Auto-Tests: Covered` to this PR?
What do you think?

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
Was checked cases from **Manual testing scenarios**:
**Before:**
:x: `setModalElement` function is called when guest checkout is enabled
#30647Main


**After:**
:heavy_check_mark: `setModalElement` function is no longer called when guest checkout is enabled
#30647PrProfiler
:heavy_check_mark: Authentication popup shows properly when guest user tries to go to the checkout but guest checkout is disabled.
Peek 2020-12-04 14-13
:heavy_check_mark: Authentication popup shows properly when guest user tries place order with downloadable product and `Disable Guest Checkout if Cart Contains Downloadable Items=Yes`
Peek 2020-12-04 14-07

**Comment by engcom-Delta** [comment]
**Note:** Automation tests are passed

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-8387](https://jira.corp.magento.com/browse/ENGCOM-8387) has been created to process this Pull Request

**Comment by gabrieldagama** [comment]
@magento run al tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
Hi, @engcom-Charlie can you check the failed tests on this PR? It seems to be constantly failing.

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by engcom-Charlie** [comment]
@magento run Functional Tests CE, Functional Tests EE, Functional Tests B2B

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed
Result is same as in [comment](https://github.com/magento/magento2/pull/30647#issuecomment-738976134)

**Comment by engcom-Delta** [comment]
**Note:** Functional Tests B2B are failed

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8387](https://jira.corp.magento.com/browse/ENGCOM-8387) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
**Note:** No new changes since last check https://github.com/magento/magento2/pull/30647#issuecomment-759028728
Automation tests are passed

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by gabrieldagama** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, WebAPI Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE, WebAPI Tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests CE, Functional Tests EE

**Review by gabrieldagama — CHANGES_REQUESTED** [review]
Hi @engcom-Foxtrot it seems this PR is failing tests constantly and has conflicts, can you have a look on it?

File: app/code/Magento/Customer/view/frontend/web/js/view/authentication-popup.js [hunk 2]
```diff
@@ -41,8 +41,27 @@ define([
...
+            }
+        },
+
+        /**
+         * Initializes authentication modal on given HTML element.
+         */
+        createPopup: function (element) {
             if (authenticationPopup.modalWindow == null) {
                 authenticationPopup.createPopUp(element);
             }
```
Comment (mrtuvn) [inline app/code/Magento/Customer/view/frontend/web/js/view/authentication-popup.js:67]: should we compare with authenticationPopup.modalWindow === null (triple equal with type convert) instead ?

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@engcom-Foxtrot  still failing,...

**Comment by mrtuvn** [comment]
Can i create seperate PR for this

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@engcom-Foxtrot can you update/fix for conflicts file

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by ihor-sviziev** [comment]
@engcom-Foxtrot tests still failing :(

**Comment by engcom-Foxtrot** [comment]
@magento run Functional Tests CE, Functional Tests CE, Functional Tests B2B

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run all tests

**Comment by engcom-Foxtrot** [comment]
@magento run Functional Tests CE, Functional Tests B2B, WebAPI Tests

**Comment by ihor-sviziev** [comment]
@magento run Functional Tests B2B

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B

**Review by ihor-sviziev — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @ihor-sviziev, thank you for the review.
[ENGCOM-8387](https://jira.corp.magento.com/browse/ENGCOM-8387) has been created to process this Pull Request

**Review by Den4ik — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @Den4ik, thank you for the review.
[ENGCOM-8387](https://jira.corp.magento.com/browse/ENGCOM-8387) has been created to process this Pull Request

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
Thank you for your contribution @krzksz. 

Manual scenarios are working as expected but one test case from Functional test CE is still failing. 

:x: QA not Passed

**Manual testing scenarios**:

**Before:**
:x: `setModalElement` function is called when guest checkout is enabled

image

**After:**
:heavy_check_mark: `setModalElement` function is no longer called when guest checkout is enabled

image

:heavy_check_mark: Authentication popup shows properly when guest user tries to go to the checkout with guest checkout is disabled.

image


https://user-images.githubusercontent.com/51680850/130239324-530d7119-5ce9-49ca-8d8d-96b34b2b0ac9.mp4

**Comment by mrtuvn** [comment]
@magento run Functional Tests B2B, Functional Tests CE

**Comment by engcom-Charlie** [comment]
Hi @krzksz, thank you for your contribution.

As per review comments and testing one of the test case from this PR which is StoreFrontCheckVatIdAtAccountCreateWithMultishipmentTest is broken.

I am working on it.

Thank you!

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
@magento run all tests

**Comment by engcom-Charlie** [comment]
Thanks for everyone's contribution!!

As per your review comments, I have merged magento2/2.4-develop into this PR branch.

I have tested all MFTF test cases which are getting covered in this PR. All are working fine before and after merge.

The above failed test cases are not part of this PR and also the PR changes are not causing the failure of above test cases so moving this PR in testing.

**Comment by ihor-sviziev** [comment]
Hi @engcom-Charlie, the Functional Tests CE and Functional Tests EE are still failing after restarting...
It looks like some issue is still there. Isn't it?

**Comment by mrtuvn** [comment]
branch seem too old for automate tasks. I think update code branch may help

**Comment by Den4ik** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run all tests

**Comment by mrtuvn** [comment]
@magento run Functional Tests CE, Functional Tests B2B

**Comment by engcom-Alfa** [comment]
:heavy_check_mark: QA Passed

**Preconditions:**

1. Have an Magento installed with sample data


**Manual testing scenario:**

1. Set the Store>Confgiration>Sales>Checkout> Checkout Options > _Allow Guest Checkout - Yes_

2. Front end> Don't customer login; just add product to cart and proceed to checkout
 
3. Validate the setModalElement even has been called or no


**Before:** :heavy_multiplication_x: _setModalElement_ event was getting called when the guest checkout is enabled

image


**After:** :heavy_check_mark:  _setModalElement_ event is  not getting called when the guest checkout is enabled

image


Additionally tested below scenarios with the PR fix code:
1. Set Guest checkout Yes: Logged in as a existing customer from the checkout page and able to place the order.
2. Set Guest checkout Yes: Just provided the shopping address with a non existing mail address from the checkout page and able to place the order.
3. Set Guest checkout No: Tried to checkout without logging in, received the popup to login first and then able to place the order.

Since this PR has the fix relevant to avoid the unused even from the request, and it has no impact on other features as such, no additional regression is required apart from above listed cases.

**Comment by ishakhsuvarov** [comment]
Moving to `extended testing` to get merge conflicts resolved.

**Comment by engcom-Echo** [comment]
@magento run all tests

**Comment by engcom-Echo** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Functional Tests EE, Functional Tests B2B

**Comment by engcom-Echo** [comment]
@magento run Functional Tests B2B

**Comment by engcom-Echo** [comment]
Seems like bot moved this PR from `Merge in Progress` to `Ready for Testing`. Moving back to `merge in progress`.

**Comment by ihor-sviziev** [comment]
Finally! This or got 2 birthdays and almost 1 month, and only then merged!

**Pull request merged**